lib/rs/src/mock/mod.rs

```diff
@@ -0,0 +1,255 @@
+use std::io;
+use std::io::Read as IoRead;
+use std::io::Write as IoWrite;
+
+use {Protocol, Transport, Result};
+use protocol::{Type, MessageType};
+
+pub use self::ProtocolAction::*;
+pub use self::Primitive::*;
+pub use self::Action::*;
+
+#[derive(Debug, Clone)]
+pub struct MockTransport {
+    reader: io::Cursor<Vec<u8>>,
+    writer: Vec<u8>,
+}
+
+impl MockTransport {
+    pub fn new(buf: Vec<u8>) -> MockTransport {
+        MockTransport {
+            reader: io::Cursor::new(buf),
+            writer: Vec::new()
+        }
+    }
+}
+
+impl io::Write for MockTransport {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.writer.write(buf)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        self.writer.flush()
+    }
+}
+
+impl io::Read for MockTransport {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        self.reader.read(buf)
+    }
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct MockProtocol {
+    log: Vec<ProtocolAction>
+}
+
+impl MockProtocol {
+    pub fn new() -> Self { Default::default() }
+    pub fn log(&self) -> &[ProtocolAction] { &self.log }
+
+    fn log_action(&mut self, action: ProtocolAction) -> Result<()> {
+        self.log.push(action);
+        Ok(())
+    }
+}
+
+macro_rules! read {
+    ($selff:expr, $expected:pat, $body:expr) => {{
+        if $selff.log.len() == 0 {
+            panic!(concat!("Unexpected read on empty log. Expected ", stringify!($expected)))
+        }
+
+        match $selff.log.remove(0) {
+             $expected => Ok($body),
+             other => {
+                 panic!(concat!("Unexpected read. Expected ", stringify!($expected),
+                        ", encountered {:?}. Log was: {:?}"), &other, &$selff.log)
+             }
+        }
+    }};
+    ($selff:expr, $expected:pat) => { read!($selff, $expected, ()) }
+}
+
+// omg
```

Comment:
haha, sorry, just complaining to myself about the extremely mechanical code that follows